### PR TITLE
Fix to remove default off detectors from experiments

### PR DESCRIPTION
**Pull request by Omotola**
PR to remove experiments for unused/disabled detectors from experiments that run. Tested locally with the NewPipDetectorExperiment. The experiment runs only when the detector is enabled as expected.

### CODE REVIEW

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs [hunk 1]
```diff
@@ -108,6 +108,24 @@ private void FilterExperiments(IComponentDetector detector, int count)
         }
     }
 
+    public void RemoveUnwantedExperimentsbyDetectors(IEnumerable<IComponentDetector> detectors)
+    {
+        var experimentsToRemove = new List<IExperimentConfiguration>();
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs:113]: This allocation isn't necessary. You can just do `var experimentsToRemove = [linq query]`

File: src/Microsoft.ComponentDetection.Orchestrator/Services/IDetectorProcessingService.cs [hunk 2]
```diff
@@ -8,4 +8,6 @@
 public interface IDetectorProcessingService
 {
     Task<DetectorProcessingResult> ProcessDetectorsAsync(IDetectionArguments detectionArguments, IEnumerable<IComponentDetector> detectors, DetectorRestrictions detectorRestrictions);
+
+    void SetUnusedDetectors(IEnumerable<IComponentDetector> detectors);
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Services/IDetectorProcessingService.cs:12]: super nit: I don't like this. Is there any way we could maybe hook into the detector restriction server service instead?

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs [hunk 3]
```diff
@@ -108,6 +108,24 @@ private void FilterExperiments(IComponentDetector detector, int count)
         }
     }
 
+    public void RemoveUnwantedExperimentsbyDetectors(IEnumerable<IComponentDetector> detectors)
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs:111]: Could you add tests for this?

File: src/Microsoft.ComponentDetection.Orchestrator/Services/IDetectorProcessingService.cs (hunk 2 above)
Comment (Omotola) [reply src/Microsoft.ComponentDetection.Orchestrator/Services/IDetectorProcessingService.cs]: update: added the disabled detectors to `DetectorRestrictions`

File: src/Microsoft.ComponentDetection.Orchestrator/Services/BcdeScanExecutionService.cs [hunk 4]
```diff
@@ -39,6 +39,7 @@ public async Task<ScanResult> ExecuteScanAsync(IDetectionArguments detectionArgu
 
         var detectorRestrictions = this.GetDetectorRestrictions(detectionArguments);
         var detectors = this.detectorRestrictionService.ApplyRestrictions(detectorRestrictions, this.detectors).ToImmutableList();
```
Comment (cobya) [inline src/Microsoft.ComponentDetection.Orchestrator/Services/BcdeScanExecutionService.cs:41]: nit: I know this isn't your variable but if we're going to be using `this.detectors` and `detectors` in the same function we should probably rename it to something like `restrictedDetectors`

File: src/Microsoft.ComponentDetection.Orchestrator/Services/DetectorProcessingService.cs [hunk 5]
```diff
@@ -349,4 +352,6 @@ private void LogTabularOutput(ILogger logger, ConcurrentDictionary<string, Detec
             this.logger.LogInformation("{Line}", line);
         }
     }
+
+    public void SetUnusedDetectors(IEnumerable<IComponentDetector> detectors) => this.unusedDetectors = detectors;
```
Comment (cobya) [inline src/Microsoft.ComponentDetection.Orchestrator/Services/DetectorProcessingService.cs:356]: Is this needed anymore? I don't see any references to unusedDetectors

**Review by cobya — APPROVED** [review]

**Pull request merged**